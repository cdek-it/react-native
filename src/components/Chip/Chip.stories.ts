import type { Meta, StoryObj } from '@storybook/react'
import { IconCheck } from '@tabler/icons-react-native'

import { Chip } from './Chip'

const meta: Meta<typeof Chip> = {
  title: 'Chip',
  component: Chip,
  args: {
    showClose: true,
    disabled: false,
  },
  argTypes: {
    showClose: { control: 'boolean' },
    disabled: { control: 'boolean' },
  },
  tags: ['autodocs'],
}
export default meta

type Story = StoryObj<typeof Chip>

const ChipStory: Story = {
  args: {
    children: 'Chip',
    Icon: IconCheck,
    showClose: true,
    disabled: false,
    onClose: () => {
      console.log('Tap close Chip')
    },
    onPress: () => {
      console.log('Tap Chip')
    },
  },
}

export { ChipStory as Chip }
