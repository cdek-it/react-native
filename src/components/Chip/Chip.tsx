import { type Icon, IconX } from '@tabler/icons-react-native'
import React from 'react'
import { Text, Pressable, type PressableProps } from 'react-native'

import { makeStyles, type StylesObject } from '../../utils/makeStyles'

export type ChipProps = Omit<PressableProps, 'disabled'> &
  React.PropsWithChildren<{
    /** Иконка из библиотеки Tabler */
    Icon?: Icon
    /**
     * Признак для отображения крестика внутри компонента
     * @default true
     */
    showClose: boolean
    /** Обработчик нажатия на крестик */
    onClose?: () => void
    /** Признак доступности для нажатия */
    disabled?: boolean
  }>

/** Компонет Нажимабельного тега с возможностью закрытия
 * Используется для представления массива данных в виде меток
 */
export const Chip = (props: ChipProps) => {
  const { Icon, showClose, children, disabled = false, onClose, ...rest } = props
  const styles = useStyles()

  return (
    <Pressable
      {...rest}
      disabled={disabled}
      style={({ pressed }) => [
        styles.chip,
        disabled && styles.disabledChip,
        pressed && styles.highlightChip,
      ]}
    >
      {Icon && (
        <Icon
          size={styles.icon.height as number}
          style={[styles.icon, disabled && styles.disabledIcon]}
        />
      )}

      <Text numberOfLines={1} style={[styles.text, disabled && styles.disabledText]}>
        {children}
      </Text>

      {showClose && (
        <IconX
          disabled={disabled}
          size={styles.icon.height as number}
          style={[styles.icon, disabled && styles.disabledIcon]}
          onPress={onClose}
        />
      )}
    </Pressable>
  )
}
const useStyles = makeStyles<StylesObject>(({ theme }) => ({
  chip: {
    height: theme.Misc.Chip.chipHeight,
    alignSelf: 'flex-start',
    justifyContent: 'center',
    alignItems: 'center',
    flexDirection: 'row',
    gap: theme.General.inlineSpacing,

    paddingHorizontal: theme.Misc.Chip.chipPaddingLeftRight,
    paddingVertical: theme.Misc.Chip.chipPaddingTopBottom,

    borderRadius: theme.Misc.Chip.chipBorderRadius,
    borderWidth: 1,

    backgroundColor: '#E5E5E5', // theme.Misc.Chip.chipBg, - с прозрачным цветом, тень работать не будет
    borderColor: theme.Misc.Chip.chipBorderColor,
  },
  disabledChip: {
    backgroundColor: theme.Button.Disabled.disabledButtonBg,
    borderColor: theme.Button.Disabled.disabledButtonBorderColor,
  },
  highlightChip: {
    shadowColor: theme.General.focusOutlineColor,
    shadowRadius: 3.5,
    shadowOffset: { width: 0, height: 0 },
    shadowOpacity: 1,
    elevation: 3,
  },
  icon: {
    width: 14,
    height: 14,
    color: theme.Misc.Chip.chipTextColor,
  },
  disabledIcon: {
    color: theme.Button.Disabled.disabledButtonTextColor,
  },
  text: {
    fontSize: 14,
    fontWeight: 400,
    verticalAlign: 'middle',
    color: theme.Misc.Chip.chipTextColor,
  },
  disabledText: {
    color: theme.General.textSecondaryColor,
  },
}))
