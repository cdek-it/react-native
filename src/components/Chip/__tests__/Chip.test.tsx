import { render } from '@testing-library/react-native'
import React from 'react'

import { Chip, type ChipProps } from '../Chip'

describe('Chip component tests', () => {
  const snapshotCases: Array<[string, ChipProps]> = [
    ['showClose: true', { showClose: true, children: 'Chip' }],
    ['showClose: false', { showClose: false, children: 'Chip' }],
    ['showClose: true, disabled: true', { showClose: true, disabled: true, children: 'Chip' }],
    ['showClose: false, disabled: true', { showClose: false, disabled: true, children: 'Chip' }],
  ]

  test.each(snapshotCases)('%s', (_, props) => {
    const renderedChip = render(<Chip {...props} />)
    expect(renderedChip.toJSON()).toMatchSnapshot()
  })
})
