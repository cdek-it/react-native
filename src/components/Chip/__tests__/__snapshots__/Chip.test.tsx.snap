// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chip component tests showClose: false 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "alignItems": "center",
        "alignSelf": "flex-start",
        "backgroundColor": "#E5E5E5",
        "borderColor": "rgba(255, 255, 255, 0.0001)",
        "borderRadius": 3.5,
        "borderWidth": 1,
        "flexDirection": "row",
        "gap": 7,
        "height": 24.5,
        "justifyContent": "center",
        "paddingHorizontal": 7,
        "paddingVertical": 0,
      },
      false,
      false,
    ]
  }
>
  <Text
    numberOfLines={1}
    style={
      [
        {
          "color": "rgba(0, 0, 0, 0.8000)",
          "fontSize": 14,
          "fontWeight": 400,
          "verticalAlign": "middle",
        },
        false,
      ]
    }
  >
    Chip
  </Text>
</View>
`;

exports[`Chip component tests showClose: false, disabled: true 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": true,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "alignItems": "center",
        "alignSelf": "flex-start",
        "backgroundColor": "#E5E5E5",
        "borderColor": "rgba(255, 255, 255, 0.0001)",
        "borderRadius": 3.5,
        "borderWidth": 1,
        "flexDirection": "row",
        "gap": 7,
        "height": 24.5,
        "justifyContent": "center",
        "paddingHorizontal": 7,
        "paddingVertical": 0,
      },
      {
        "backgroundColor": "rgba(0, 0, 0, 0.1000)",
        "borderColor": "#dbdada",
      },
      false,
    ]
  }
>
  <Text
    numberOfLines={1}
    style={
      [
        {
          "color": "rgba(0, 0, 0, 0.8000)",
          "fontSize": 14,
          "fontWeight": 400,
          "verticalAlign": "middle",
        },
        {
          "color": "rgba(0, 0, 0, 0.6000)",
        },
      ]
    }
  >
    Chip
  </Text>
</View>
`;

exports[`Chip component tests showClose: true 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "alignItems": "center",
        "alignSelf": "flex-start",
        "backgroundColor": "#E5E5E5",
        "borderColor": "rgba(255, 255, 255, 0.0001)",
        "borderRadius": 3.5,
        "borderWidth": 1,
        "flexDirection": "row",
        "gap": 7,
        "height": 24.5,
        "justifyContent": "center",
        "paddingHorizontal": 7,
        "paddingVertical": 0,
      },
      false,
      false,
    ]
  }
>
  <Text
    numberOfLines={1}
    style={
      [
        {
          "color": "rgba(0, 0, 0, 0.8000)",
          "fontSize": 14,
          "fontWeight": 400,
          "verticalAlign": "middle",
        },
        false,
      ]
    }
  >
    Chip
  </Text>
  <RNSVGSvgView
    align="xMidYMid"
    bbHeight={14}
    bbWidth={14}
    disabled={false}
    fill="none"
    focusable={false}
    height={14}
    meetOrSlice={0}
    minX={0}
    minY={0}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    responsible={true}
    stroke="currentColor"
    strokeLinecap="round"
    strokeLinejoin="round"
    strokeWidth={2}
    style={
      [
        {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        [
          {
            "color": "rgba(0, 0, 0, 0.8000)",
            "height": 14,
            "width": 14,
          },
          false,
        ],
        {
          "flex": 0,
          "height": 14,
          "width": 14,
        },
      ]
    }
    tintColor="rgba(0, 0, 0, 0.8000)"
    vbHeight={24}
    vbWidth={24}
    width={14}
    xmlns="http://www.w3.org/2000/svg"
  >
    <RNSVGGroup
      color="rgba(0, 0, 0, 0.8000)"
      fill={null}
      propList={
        [
          "fill",
          "stroke",
          "strokeWidth",
          "strokeLinecap",
          "strokeLinejoin",
        ]
      }
      stroke={
        {
          "type": 2,
        }
      }
      strokeLinecap={1}
      strokeLinejoin={1}
      strokeWidth={2}
    >
      <RNSVGPath
        color="rgba(0, 0, 0, 0.8000)"
        d="M18 6l-12 12"
        fill={null}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        propList={
          [
            "fill",
            "stroke",
            "strokeWidth",
            "strokeLinecap",
            "strokeLinejoin",
          ]
        }
        responsible={true}
        stroke={
          {
            "type": 2,
          }
        }
        strokeLinecap={1}
        strokeLinejoin={1}
        strokeWidth={2}
      />
      <RNSVGPath
        color="rgba(0, 0, 0, 0.8000)"
        d="M6 6l12 12"
        fill={null}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        propList={
          [
            "fill",
            "stroke",
            "strokeWidth",
            "strokeLinecap",
            "strokeLinejoin",
          ]
        }
        responsible={true}
        stroke={
          {
            "type": 2,
          }
        }
        strokeLinecap={1}
        strokeLinejoin={1}
        strokeWidth={2}
      />
    </RNSVGGroup>
  </RNSVGSvgView>
</View>
`;

exports[`Chip component tests showClose: true, disabled: true 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": true,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "alignItems": "center",
        "alignSelf": "flex-start",
        "backgroundColor": "#E5E5E5",
        "borderColor": "rgba(255, 255, 255, 0.0001)",
        "borderRadius": 3.5,
        "borderWidth": 1,
        "flexDirection": "row",
        "gap": 7,
        "height": 24.5,
        "justifyContent": "center",
        "paddingHorizontal": 7,
        "paddingVertical": 0,
      },
      {
        "backgroundColor": "rgba(0, 0, 0, 0.1000)",
        "borderColor": "#dbdada",
      },
      false,
    ]
  }
>
  <Text
    numberOfLines={1}
    style={
      [
        {
          "color": "rgba(0, 0, 0, 0.8000)",
          "fontSize": 14,
          "fontWeight": 400,
          "verticalAlign": "middle",
        },
        {
          "color": "rgba(0, 0, 0, 0.6000)",
        },
      ]
    }
  >
    Chip
  </Text>
  <RNSVGSvgView
    align="xMidYMid"
    bbHeight={14}
    bbWidth={14}
    disabled={true}
    fill="none"
    focusable={false}
    height={14}
    meetOrSlice={0}
    minX={0}
    minY={0}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    responsible={true}
    stroke="currentColor"
    strokeLinecap="round"
    strokeLinejoin="round"
    strokeWidth={2}
    style={
      [
        {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        [
          {
            "color": "rgba(0, 0, 0, 0.8000)",
            "height": 14,
            "width": 14,
          },
          {
            "color": "rgba(0, 0, 0, 0.6000)",
          },
        ],
        {
          "flex": 0,
          "height": 14,
          "width": 14,
        },
      ]
    }
    tintColor="rgba(0, 0, 0, 0.6000)"
    vbHeight={24}
    vbWidth={24}
    width={14}
    xmlns="http://www.w3.org/2000/svg"
  >
    <RNSVGGroup
      color="rgba(0, 0, 0, 0.6000)"
      fill={null}
      propList={
        [
          "fill",
          "stroke",
          "strokeWidth",
          "strokeLinecap",
          "strokeLinejoin",
        ]
      }
      stroke={
        {
          "type": 2,
        }
      }
      strokeLinecap={1}
      strokeLinejoin={1}
      strokeWidth={2}
    >
      <RNSVGPath
        color="rgba(0, 0, 0, 0.6000)"
        d="M18 6l-12 12"
        fill={null}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        propList={
          [
            "fill",
            "stroke",
            "strokeWidth",
            "strokeLinecap",
            "strokeLinejoin",
          ]
        }
        responsible={true}
        stroke={
          {
            "type": 2,
          }
        }
        strokeLinecap={1}
        strokeLinejoin={1}
        strokeWidth={2}
      />
      <RNSVGPath
        color="rgba(0, 0, 0, 0.6000)"
        d="M6 6l12 12"
        fill={null}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        propList={
          [
            "fill",
            "stroke",
            "strokeWidth",
            "strokeLinecap",
            "strokeLinejoin",
          ]
        }
        responsible={true}
        stroke={
          {
            "type": 2,
          }
        }
        strokeLinecap={1}
        strokeLinejoin={1}
        strokeWidth={2}
      />
    </RNSVGGroup>
  </RNSVGSvgView>
</View>
`;
